File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Panama.m
//
//  UIImage+Panama.m
//  CountriesFlags
//

#import "Countries/UIImage+Panama.h"

#import "Extensions/UIBezierPath+Star.h"

@implementation UIImage (Panama)

+ (UIImage * _Nullable)flagPanamaWithSize:(CGSize)size {
    UIColor* colorBlue = [UIColor colorWithRed:1.0f/255.0f
                                         green:30.0f/255.0f
                                          blue:86.0f/255.0f
                                         alpha:1.0f];
    UIColor* colorRed = [UIColor colorWithRed:219.0f/255.0f
                                        green:10.0f/255.0f
                                         blue:19.0f/255.0f
                                        alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    
    UIGraphicsBeginImageContextWithOptions(size, NO, 0.0f);
    
    [colorWhite setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorRed setFill];
    UIRectFill(CGRectMake(size.width / 2.0f, 0.0f, size.width / 2.0f, size.height / 2.0f));
    
    [colorBlue setFill];
    UIRectFill(CGRectMake(0.0f, size.height / 2.0f, size.width / 2.0f, size.height / 2.0f));
    
    CGFloat radius = size.height / 8.0f * 1.111145618;
    CGFloat xDelta = size.height / 8.0f * 0.105572809;
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 4.0f, (size.height / 2.0f + xDelta * 2.0f) / 2.0f)
                        radius:radius
                          type:FlagStarTypeFivePointed] fill];
    [colorRed setFill];
    [[UIBezierPath starAtPoint:CGPointMake(size.width * 0.75f, (size.height * 1.5f + xDelta * 2.0f) / 2.0f)
                        radius:radius
                          type:FlagStarTypeFivePointed] fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Panama.swift
//
//  UIImage+Panama.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Panama "wikipedia"
     
     Generates the flag of Panama.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Panama.
     */
    internal static func flagPanama(with size: CGSize) -> UIImage? {
        let colorBlue = UIColor(red: 1.0 / 255.0,
                                green: 30.0 / 255.0,
                                blue: 86.0 / 255.0,
                                alpha: 1.0)
        let colorRed = UIColor(red: 219.0 / 255.0,
                               green: 10.0 / 255.0,
                               blue: 19.0 / 255.0,
                               alpha: 1.0)
        let colorWhite = UIColor.white

        UIGraphicsBeginImageContextWithOptions(size, false, 0.0)

        colorWhite.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorRed.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: size.width / 2.0, y: 0.0),
                          size: size.halfSize))

        colorBlue.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: 0.0, y: size.height / 2.0),
                          size: size.halfSize))

        let radius = size.height / 8.0 * 1.111145618
        let xDelta = size.height / 8.0 * 0.105572809
        UIBezierPath.star(at: CGPoint(x: size.width / 4.0, y: (size.height / 2.0 + xDelta * 2.0) / 2.0),
                          radius: radius).fill()
        colorRed.setFill()
        UIBezierPath.star(at: CGPoint(x: size.width * 0.75, y: (size.height * 1.5 + xDelta * 2.0) / 2.0),
                          radius: radius).fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
